rclpy/test/test_client.py

```diff
@@ -90,6 +90,25 @@ def test_concurrent_calls_to_service(self):
             self.node.destroy_client(cli)
             self.node.destroy_service(srv)
 
+    def test_different_type_raises(self):
+        cli = self.node.create_client(GetParameters, 'get/parameters')
+        srv = self.node.create_service(
+            GetParameters, 'get/parameters',
+            lambda request, response: 'different response type')
+        try:
+            with self.assertRaises(TypeError):
+                cli.call('different request type')
+            with self.assertRaises(TypeError):
+                cli.call_async('different request type')
+            self.assertTrue(cli.wait_for_service(timeout_sec=20))
+            future = cli.call_async(GetParameters.Request())
+            executor = rclpy.executors.SingleThreadedExecutor(context=self.context)
+            with self.assertRaises(TypeError):
+                rclpy.spin_until_future_complete(self.node, future, executor=executor)
```

Comment:
What does this traceback look like? I would have expected the exception to be swallowed here:

https://github.com/ros2/rclpy/blob/afc4cd407143b1b019fa2306403b7584a349b97a/rclpy/rclpy/task.py#L212-L213

Reply:
Output of the test without the `with self.assertRaises(TypeError):`:

```
rclpy/__init__.py:204: in spin_until_future_complete
    executor.spin_until_future_complete(future, timeout_sec)
rclpy/executors.py:271: in spin_until_future_complete
    self.spin_once(timeout_sec=timeout_sec)
rclpy/executors.py:628: in spin_once
    raise handler.exception()
rclpy/task.py:206: in __call__
    self._handler.send(None)
rclpy/executors.py:382: in handler
    await call_coroutine(entity, arg)
rclpy/executors.py:337: in _execute_service
    srv.send_response(response, header)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <rclpy.service.Service object at 0x7f549dd3bc88>
response = 'different response type'
header = <capsule object "rmw_request_id_t" at 0x7f549dd3c4e0>

    def send_response(self, response: SrvTypeResponse, header) -> None:
        """
        Send a service response.
    
        :param response: The service response.
        :param header: Capsule pointing to the service header from the original request.
        :raises: TypeError if the type of the passed response isn't an instance
          of the Response type of the provided service when the service was
          constructed.
        """
        if not isinstance(response, self.srv_type.Response):
>           raise TypeError()
E           TypeError

rclpy/service.py:79: TypeError
```

Reply:
Ohh I didn't know about https://github.com/ros2/rclpy/blob/afc4cd407143b1b019fa2306403b7584a349b97a/rclpy/rclpy/executors.py#L627-L628 That's a nice improvement